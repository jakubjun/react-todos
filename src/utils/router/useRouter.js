import useLocation from './useLocation';
import configRoutes from '../../config/routes';

export const notFoundPathName = 'notfound';

export default function useRouter(routes = configRoutes) {
  const path = useLocation();

  return routes[path] || routes[notFoundPathName];
}
