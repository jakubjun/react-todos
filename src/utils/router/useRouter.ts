import { FC } from 'react';
import useLocation from './useLocation';

export type Route = [string, FC<any>, FC<any>];

export type NotFoundRoute = [FC<any>, FC<any>];

export default function useRouter(
  routes:Route[],
  notFoundRoute: [FC<any>, FC<any>],
): NotFoundRoute {
  const path = useLocation();

  const matchedRoute = routes.find(([pattern]) => {
    if (typeof pattern === 'string') {
      return pattern === path;
    }

    // TODO add support for regex route patterns
    // if (pattern instanceof RegExp) {
    //   return pattern.test(path);
    // }

    return false;
  });

  if (!matchedRoute) {
    return notFoundRoute;
  }

  return [matchedRoute[1], matchedRoute[2]];
}
