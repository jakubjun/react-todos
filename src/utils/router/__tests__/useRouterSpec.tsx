import renderer from 'react-test-renderer';
import React from 'react';
import useRouter, { NotFoundRoute, Route } from '../useRouter';

const routesConfig:Route[] = [
  ['/', () => <div>index layout</div>, () => <div>index page</div>],

];

const notFoundRoute:NotFoundRoute = [() => <div>index layout</div>, () => <div>index page</div>];

const cases = [
  '/',
];

describe('useRouter', () => {
  beforeEach(() => {
    // @ts-ignore
    delete window.location;
  });

  test.each(cases)('should route to %s', (route) => {
    window.location = new URL(`http://hello.world${route}`) as any;
    const [Layout, Page] = useRouter(routesConfig, notFoundRoute);
    const layoutComponent = renderer.create(<Layout />).toJSON();
    const pageComponent = renderer.create(<Page />).toJSON();
    expect(layoutComponent).toMatchSnapshot();
    expect(pageComponent).toMatchSnapshot();
  });
});
