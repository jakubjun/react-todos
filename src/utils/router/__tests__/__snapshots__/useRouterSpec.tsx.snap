// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRouter should route to / 1`] = `
<div>
  index layout
</div>
`;

exports[`useRouter should route to / 2`] = `
<div>
  index page
</div>
`;
